### PR TITLE
Support multiple APIs in one app

### PR DESCRIPTION
**Pull request by anatol-sialitski**
#10514

### CODE REVIEW

File: modules/core/core-api/src/main/java/com/enonic/xp/api/ApiDescriptor.java [hunk 1]
```diff
@@ -35,9 +33,21 @@ public boolean isAccessAllowed( final PrincipalKeys principalKeys )
...
     {
-        return ResourceKey.from( applicationKey, API_DESCRIPTOR_PATH );
+        return toResourceKey( this.key, extension );
+    }
+
+    public static ResourceKey toResourceKey( final DescriptorKey key, final String extension )
+    {
+        if ( "api".equals( key.getName() ) )
+        {
+            return ResourceKey.from( key.getApplicationKey(), "api/api." + extension );
```
Comment (rymsha) [inline modules/core/core-api/src/main/java/com/enonic/xp/api/ApiDescriptor.java:45]:
```suggestion
            return ResourceKey.from( key.getApplicationKey(), "apis/api." + extension );
```

File: modules/core/core-api/src/main/java/com/enonic/xp/api/ApiDescriptor.java [hunk 2]
```diff
@@ -35,9 +33,21 @@ public boolean isAccessAllowed( final PrincipalKeys principalKeys )
...
+        }
+        else
+        {
+            return ResourceKey.from( key.getApplicationKey(), "api/" + key.getName() + "/" + key.getName() + "." + extension );
```
Comment (rymsha) [inline modules/core/core-api/src/main/java/com/enonic/xp/api/ApiDescriptor.java:49]:
```suggestion
            return ResourceKey.from( key.getApplicationKey(), "apis/" + key.getName() + "/" + key.getName() + "." + extension );
```

File: modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/api/ApiDescriptorKeyLocator.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+import java.util.Set;
+
+import com.google.common.collect.ImmutableSet;
+
+import com.enonic.xp.app.ApplicationKey;
+import com.enonic.xp.page.DescriptorKey;
+import com.enonic.xp.resource.ResourceService;
+
+public class ApiDescriptorKeyLocator
```
Comment (rymsha) [inline modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/api/ApiDescriptorKeyLocator.java:11]: why DescriptorKeyLocator is not sufficient?

File: modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/handler/api/ApiAppHandler.java [hunk 4]
```diff
@@ -37,7 +38,8 @@
 public class ApiAppHandler
     extends BaseWebHandler
 {
-    public static final Pattern PATTERN = Pattern.compile( "^/api/(?<appKey>(?!media|idprovider)[^/]+)/(?<restPath>.*)?$" );
+    public static final Pattern PATTERN =
+        Pattern.compile( "^/api/(?<appKey>(?!media|idprovider)[^/]+)/(?<apiKey>[^/]+)/(?<restPath>.*)?$" );
```
Comment (rymsha) [inline modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/handler/api/ApiAppHandler.java:42]: idprovider is no longer in /api

File: modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/api/ApiDescriptorKeyLocator.java (hunk 3 above)
Comment (anatol-sialitski) [reply modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/api/ApiDescriptorKeyLocator.java:11]: We must support both, for example, `apis/api.xml` and `apis/myapi/myapi.xml`.

File: modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/handler/api/ApiAppHandler.java (hunk 4 above)
Comment (anatol-sialitski) [reply modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/handler/api/ApiAppHandler.java:42]: Yes, but that is within `(?!...)` in other word it must not match. I think that part will be extended the rest of reserved words.

**Review by rymsha — APPROVED** [review]

**Pull request merged**